### PR TITLE
Fix #112 and add test

### LINKED ISSUE #112: url() incorrectly detects beginning of query string by decoding request_uri [rt.cpan.org #83265]
https://rt.cpan.org/Ticket/Display.html?id=83265

```
My CGI script gets a request such as:

GET /~rrt/Software/DarkGlass/Why%20DarkGlass%3F HTTP/1.1

Note the "%3F" at the end of the URL: it's an escaped question mark.

My script then asks for the URL using "url()", and gets:

"Software/DarkGlass/Why DarkGlass"

The question mark has been stripped off, because after unescaping in the
line:

    my $request_uri =  unescape($self->request_uri) || '';

the code then strips the query off:

    my $uri         =  $rewrite && $request_uri ? $request_uri :
$script_name;
    $uri            =~ s/\?.*$//s;                                #
remove query string

But this is wrong: if a query string is passed, it uses an unescaped
question mark. The "unescape" routine is itself undocumented, being part
of CGI::Util which is for internal use only. I'm not an expert, so I
don't know if some unescaping should be being done, but it seems that at
the very least, question marks should not be unescaped at that point.
However, since the question mark denoting a query string should be a
literal question mark character in the original request URI, it seems to
be that it should be safe to remove the query string before unescaping
the URI.

The code in question has not been changed since the start of CGI.pm's
git history, so I can't find any relevant commit logs.

If I'm correct, I'm still not exactly sure what a minimum patch would
be, as while the obvious thing to do would be to move the call of
unescape to after the stripping of the query string, that would mean
applying unescape to the value of $script_name in some circumstances,
which might be OK, or might not, I don't know. But I guess $script_name
shouldn't contain a query string, so in fact, stripping the query string
should only be needed in the case that $uri is set to $request_uri, so
it should be possible to rewrite the relevant lines from:

    my $request_uri =  unescape($self->request_uri) || '';
    my $query_str   =  $self->query_string;

    my $rewrite_in_use = $request_uri && $request_uri !~ /^\Q$script_name/;

    my $uri         =  $rewrite && $request_uri ? $request_uri :
$script_name;
    $uri            =~ s/\?.*$//s;                                #
remove query string

to:

    my $request_uri =  $self->request_uri || '';
    my $query_str   =  $self->query_string;

    $request_uri            =~ s/\?.*$//s;                             #
remove query string
    $request_uri = unescape($request_uri);
    my $rewrite_in_use = $request_uri && $request_uri !~ /^\Q$script_name/;

    my $uri         =  $rewrite && $request_uri ? $request_uri:
$script_name;
```

### PR DESCRIPTION
**Pull request by rrthomas**

### CODE REVIEW

File: lib/CGI.pm [hunk 1]
```diff
@@ -931,7 +931,7 @@ sub _setup_symbols {
 
 	if ( /^[:-]any$/ ) {
 		warn "CGI -any pragma has been REMOVED. You should audit your code for any use "
-			. "of none supported / incorrectly spelled tags and remove them"
+			. "of unsupported/incorrectly-spelled tags and remove them"
```
Comment (leejo) [inline lib/CGI.pm]: needless churn, please remove

Comment (rrthomas) [reply lib/CGI.pm]: This is not needless: "none supported" is incorrect English, and it's not obvious what it means. But as I said, I'm happy to put it in another pull request, so I'll do that.

Comment (leejo) [reply lib/CGI.pm]: In the context of the error message there is no ambiguity in what "none supported" means here.

Comment (rrthomas) [reply lib/CGI.pm]: I've moved this fix to PR #223, and updated this PR.

**Comment by leejo** [comment]
Looks good, thanks! However, please could you write a better commit message?

**Comment by rrthomas** [comment]
I have put more detail into the commit message.

**Comment by leejo** [comment]
Excellent, thanks! Looks like 5.8 is being difficult in CI due to the coverage run, i'll fix that after merge.

**Comment by leejo** [comment]
I won't be able to release this for a couple of weeks as I am going away on vacation tomorrow, and it would be irresponsible of me to upload a release and then disappear for a while. So I'll aim to get a release out when I return, in early November.

**Comment by rrthomas** [comment]
No hurry here, I won't be getting it until it makes an Ubuntu LTS release in any case. Thanks very much for getting the fix in.

**Comment by leejo** [comment]
OK, 4.37 has been uploaded to CPAN. Thanks again!

**Comment by rrthomas** [comment]
Great stuff, that gives me a version number to add to my own code so I know when I can remove the workaround. Thanks for your efforts in getting a fix into this basically-frozen code!

**Pull request merged**